### DISCUSSION TITLE
Supervisor isolated physical network access/macvlan support

### DISCUSSION THREAD

**Discussion by agners** [home-assistant/architecture#1034] [General]
# Problem statement

Add-ons (in Supervisor based installations, so Home Assistant OS or Supervised) run by default in isolated network environment in a common, internal network named `hassio`. Each add-on gets its own internal IP address and applications can reach the user's local network and the internet via NAT.

Optionally, add-on developers can request host network access via the `host_network` [add-on configuration option](https://developers.home-assistant.io/docs/add-ons/configuration#optional-configuration-options).

The isolation provided by these two options are in rather stark contrast: On one hand the default networking isolates the add-on to a large degree from the user's local network, on the other hand using `host_network` option has no isolation at all and gives add-on full access to the hosts network interfaces, along with management interfaces like the `hassio` network itself, which is often not of interest for the add-ons and adds to confusion and issues at times.

Over time, various advanced networking needs have arisen by different type of add-ons:
- Many add-ons would like to make use of mDNS hostname resolution. This has largely been resolved: Add-on can use the regular DNS resolver through a locally run DNS (DNS plug-in) which resolves `.local` addresses as well.
- Many add-ons would like to resolve services through mDNS/DNS-SD. The [Multicast plug-in's](https://github.com/home-assistant/plugin-multicast/tree/master) [mdns-repeater](https://github.com/pvizeli/mdns-repeater) service forwards the link-local mDNS multicasts on the host network to the hassio network, which allows mDNS clients on the hassio network to learn about services on the outside network.
- Custom multicast protocol. E.g. the RaspberryMatic add-on would like to listen to custom multicast protocol (see https://github.com/jens-maus/RaspberryMatic/issues/1373). Previous attempts to use multicast routing via [pimd have not worked out](https://github.com/home-assistant/plugin-multicast/pull/61).
- IPv6 support in general. Currently the isolated network supports IPv4 only. Technically IPv6 could be supported in a similar fashion to how IPv4 is supported (using NAT), this is not possible for all IPv6 use cases (and also not how IPv6 is run, ideally). 
- IPv6 router support. The OpenThread Border Router add-on creates a separate network interface using the Linux TUN driver and acts as Thread border router routing IPv6 packets. The OTBR uses the IPv6 Neighbor Discovery Protocol to announce itself on the local network and find other OTBRs. The OTBR is a rather specific use case, but it could potentially also benefit from such a feature.

Almost all use-cases can (and typically are) solved by simply using host network, however, with the downside of having no isolation and higher chance of add-ons interfering with each other.

Another networking need comes from multi-homed hosts: Often users would like to offer services only for a particular network (interface). E.g. the DHCP add-on allows to configure the interface it hands out leases to. The Samba add-on listens on all interfaces by default, but that might not be ideal for all use cases. The OpenThread Border Router add-on uses the primary interface as deemed by the NetworkManager as backbone/infrastructure link. 

While most services allow to define the network interface they listen on, some listen on any interface by default. E.g. the Matter Server listens to mDNS/DNS-SD announcements on any interface without a built-in option to limit the interfaces used. Since this is mostly a passive listen-only, this is not that problematic. However, currently this behavior creates some log noise since the host network has also virtual interfaces from all containers, which do not have IPv6 addressing enabled:

```
2024-02-02 15:47:51 core-matter-server chip.DL[224] ERROR MDNS failed to join multicast group on veth7b5ad0c for address type IPv4: src/inet/UDPEndPointImplSockets.cpp:777: Inet Error 0x00000110: Address not found
2024-02-02 15:47:51 core-matter-server chip.DL[224] ERROR MDNS failed to join multicast group on veth912a60c for address type IPv4: src/inet/UDPEndPointImplSockets.cpp:777: Inet Error 0x00000110: Address not found
```

This architecture discussion aims to discuss these needs and find a better solution which fulfills many of the above mentioned requirements. The goal is to strike a good balance between capability, usability and usefulness of such a feature. 

# Possible solution

## Add macvlan support to Supervisor

One possible solution could be to support macvlan support to Supervisor. Docker offers the `macvlan` network driver which allows to add a virtual network interface with a separate MAC address which is bridged to the physical network interface. Since this is a separate L2 virtual network interface, it needs its own IP configuration. By default static IP configuration is available from Docker.

Implementing `macvlan` support could be done in various ways in HA: In a simple version we could just add a "virtual network" configuration page in the add-on, and allow the user to manually configure a macvlan interface. A more advanced version of this would allow multiple macvlan interfaces per add-on.

Even more advanced would be a common place to create a named "virtual network" which would allow to add multiple macvlan interfaces along with the network configurations for each one of those. These then could be attached to add-ons.

Downsides:
- Additional IP address for the isolated add-ons
- Added complexity
- Promiscuous mode support required (is this available and safe to be used on all hardware :thinking: ?)

Upsides:
- Strong isolation possible
- Flexible approach
- Full multicast support
- Transparent for applications (can be applied generically)
- Can also provide generic interface selection

**Comment by agners** [comment]
Continuing discussions from https://github.com/home-assistant/addons/pull/3416#issuecomment-1900904548

> Creating macvlan networks could go on the System -> Network page ... But where would the associations to add-ons be configured? (Maybe in each add-on's Info or Configuration tab?) ... And how would we determine/configure which add-ons should use this vs which add-ons should use the host network (I imagine some add-ons will need to be able to access multiple physical networks, so macvlan wouldn't work for those) - Should the add-on specify whether host network or macvlan should be used, or should that be configurable by the user for all add-ons, or ...?

@PaulSD these suggestions pretty much match with what I had in mind. One thing we need to keep in mind is how do we deal with existing add-ons and how can we minimize the user impact for people which do not need/want this feature.

At first I thought this could be an add-on configuration: An add-on opt-in to `macvlan` support. If enabled, the user must configure a macvlan for it.

But thinking a bit more about it, I start to feel that this should be an alternative option for `host_network`. E.g. by default the add-on with `host_network` just runs in host network. But the user optionally can assign such an add-on a `macvlan`. So existing add-ons would just continue to work transparently. And only users who want a defined set of interface exposed to a add-on would need to go through the work of setting up a macvlan :thinking: Furthermore, this would also have the least chance of negative impact, e.g. in case promiscuous mode doesn't work on some hardware or other issues with the feature.

**Comment by agners** [comment]
Continuing discussion from https://github.com/home-assistant/addons/pull/3416#issuecomment-1902463783

> I'm not sure that using macvlan is the correct solution here, since it's not that a multi-homed HA needs multiple MACs on each interface, but rather that add-ons should not make assumptions about what interfaces to use where it's necessary or desirable to pick one. Definitely some add-ons might benefit from their own L2 MAC and associated IP, but I don't think that's the main use case, and it potentially requires assigning many more external IPs to HA than necessary. I can see the benefits to assigning add-ons distinct internal IPs to permit inter-add-on communication, but not so much for external use.

Far point, one could argue macvlan is not the right tool/is over-engineered for that problem at hand. I guess my point here is if we build feature, presumably mainly for other reasons than simply network interface selection, we *could* use it as a generalized way of providing interface selection to add-ons. :thinking: I guess distinct IPs for external use is more a necessary side effect of the isolation, than really something I am looking for here. But it might also come in handy when debugging, e.g. it would make it easy to monitor the traffic of a add-on with host network access by simply filtering by IP.

> I think something inspired by [swconfig](https://openwrt.org/docs/techref/swconfig) or [DSA](https://www.kernel.org/doc/html/latest/networking/dsa/dsa.html) to assign interfaces to add-ons makes more sense than just being able to give add-ons their own macvlan(s). (It would still be possible to create an external macvlan interface and assign it to a single add-on.) As an example use case, consider a dnsmasq add-on that serves DHCP clients on some interfaces, and uses a separate upstream interface for DNS. This could also solve the issue of Matter knowing which interface to used when passed a link-local IPv6 address; it could also be leveraged to reduce broadcast traffic to interfaces that do not need it (e.g.: mDNS requests for IoT devices being sent on a non-IoT VLAN).

I am not very familiar with swconfig or DSA, but isn't this for actual switch hardware? I don't think this can be used on arbitrary hardware.

**Comment by codyc1515** [comment]
I would not support this, unless it was touchless like today. Adding IPv6 for internal traffic is a great idea but not at the expense of the user experience.

**Reply by agners** [reply]
What does "this" refer to here?

But yeah I agree, the typical use-case must not require any additional user interference. Either it is not active in that case, or things happen automatically.

**Comment by Jc2k** [comment]
I think Macvlan is a very useful tool to have in the bag. I run my HA this way, and I much prefer it to host network mode.

One of the reasons we need to carry a patch in HAOS is because running something like OTBR and matter server together in the same network namespace is sort of undefined behaviour under RFC 4191. According to the kernel maintainers, routers are not supposed to have the neighbour validity checking the matter server needs to not have 30 minute outages all the time. If we could run OTBR under Macvlan, I think we might not hit that edge case in RFC 4191 and we wouldn't need that patch?

I also wonder whether the TREL route loop that needed an OTBR patch would have behaved differently under that kind of setup?

**Comment by jens-maus** [comment]
Seeing that since Feb there haven't been any more activity here. So what is the current state of affairs here on the macvlan integration in HA? 

To make my own point a bit more clear, since I am one of the HA add-on developers which would highly benefit from being able to utilize macvlan functionality, I want to point out again that from a technical point of view a per-addon macvlan already works in HA. Just have a look at the support patch/script I have created for my users (cf. https://github.com/jens-maus/RaspberryMatic/wiki/Installation-HomeAssistant#hmip-haphmipw-drap-support-patch and https://github.com/jens-maus/RaspberryMatic/blob/master/scripts/patch-ha-addon-macvlan.sh). Executed on a HA host, this simple bash script generates the necessary docker networks (https://github.com/jens-maus/RaspberryMatic/blob/master/scripts/patch-ha-addon-macvlan.sh#L90-L94) and then simply assigns a static IP and connects the add-on docker container to this new macvlan based docker network.

So from my point of view, implementing add-on based macvlan support could/should be implemented in a rather simple and add-on centric way:

- add an additional `config.yaml` option so that add-on developers can specify that their add-on requires this `macvlan` network functionality.
- if specified allow users to enable macvlan (keep default to non macvlan) and if enabled allow to specify the static network settings (IP, subnet, etc.) in a new configuration tab in the add-on settings UI.
- if the add-on is started with enabled macvlan HA should then create the necessary docker network similar to what my current workaround support patch script is performing and connect the add-on docker container to this new network.

Therefore, I think that a general or more broad/global macvlan functionality is not really required in HA itself. IMHO add-on developers like me should know best if their add-on requires this special functionality. Thus, a simple `config.yaml` addon config option should be enough for add-on developers to allow to specify that this addon might require macvlan functionality.

So looking forward to seeing something similar being integrated in one of the next HA/supervisor versions so that my users (who require this functionality) can finally get rid of this workaround support patch which they have to re-execute on each add-on restart or HA restart. :)

**Reply by dn-gt** [reply]
Thanks for pushing this, Jens. I'm still not giving up hope that some time macvlan functionality can be enabled in Home Assistant. Currently having limitations in my setup as I'm using Rasberrymatic and cannot teach in the HM IP Wired Drap without the support patch mentioned by Jens.

**Comment by bf8392** [comment]
+1 for macvlan

**Comment by Schlichi83** [comment]
Please add support for mtacvlan in HA supervisor

**Comment by AlexCloudDev** [comment]
+1 for macvlan

**Comment by bf8392** [comment]
+1 for macvlan support in HA

**Comment by megavolts** [comment]
I know I am late, but I will be in favor of macvlan support, especially for the use case of running npm, bifrost, OTBR/matter with a dedicated IP rather than tied with HA.

+1 for macvlan support in HA

**Reply by FuNK3Y** [reply]
I am interested in macvlan as well; but I doubt that it would work properly for matter (see **https://github.com/moby/moby/issues/51447**). TBR advertises routes to thread networks, and from my preliminary testing, those routes are not taken into account by macvlan networks (whereas it works fine for host).

**Reply by agners** [reply]
@FuNK3Y what OS did you test on? I've investigated a bit into RIO processing in https://github.com/home-assistant/operating-system/discussions/2333, it might give you some pointers.

**Reply by FuNK3Y** [reply]
I did my test with ubuntu 25.10 as the host - and a ubuntu:laster / archlinux:latest.

The routes are published on the host - but not on the containers. So my kernel should have all the right options and the right sysctl flags

**Reply by FuNK3Y** [reply]
@agners - all my previous comments can be ignored. I was able to get it working with ipvlan ✌️

**Comment by agners** [comment]
Today I took a stab at this, and I was surprised how much "just worked" :sunglasses: 

<img width="1051" height="568" alt="image" src="https://github.com/user-attachments/assets/7d24453b-5984-452c-adff-1ec4a1a612ec" />

https://github.com/home-assistant/supervisor/pull/6937
https://github.com/home-assistant/frontend/pull/52572

**Reply by hmax-72** [reply]
@agners In the screenshot it's mentioned that ".... Your Home Assistant system and the app cannot reach each other ....." which indicates (from my point of view), if I'm installing Homematic, HASS would not be able to communicate with the instance.
Which App did you install? If it's Homematic, are you using HM-Wired-Components (with DRAP)?

**Reply by jens-maus** [reply]
@agners
> Today I took a stab at this, and I was surprised how much "just worked" 😎

I really, really appreciate your efforts in finally getting this MacVlan functionality done! Please note that while I was waiting for you, HA devs guys implementing this macvlan possibility, I generated a "Helper App" which is available through the usual OpenCCU HA app repository and which, in fact, applies the necessary Docker changes (MacVlan network setup, etc.) in a regular fashion, allowing even to restart the normal OpenCCU container. See here: https://github.com/OpenCCU/OpenCCU/wiki/Installation-HomeAssistant#hmiphaphmipwdrap-support-patch

However, if this new "Isolated network access" possibility will be broadly available, I could, of course, then remove that Helper App and change the documentation to assist users to enable that functionality if a HmIP-HAP or HmIPW-DRAP connection is required.

One question, though: So do I get it right that this new functionality will be available to all Apps automatically or might there even be some App config.yaml setting to tell HA that this MacVlan functionality is desired/wanted? Because perhaps it would be beneficial if an app can somewhat define to enable/disable/allow/disallow such MacVlan functionality. Or what is the usual use case here? Because if users then to enable that functionality for all kinds of Apps not even requiring this functionality, things can get IMHO quite messy and even insecure, given that not all Apps come with their own firewall functionality preventing all kinds of potentially malicious remote access, since they usually assume that the Docker isolation takes care of most access scenarios.

**Reply by mdegat01** [reply]
> One question, though: So do I get it right that this new functionality will be available to all Apps automatically or might there even be some App config.yaml setting to tell HA that this MacVlan functionality is desired/wanted?

This does seem ideal yea. Because if we go with the approach shown above it would work like this I think:
1. App config uses `host_network: true`
2. App uses host network by default on install. It works but it has all the underlying issues mentioned above
3. User is presented with the UI shown above to move it off host network and into isolation. Assuming the default works then realistically most users will not complete this step unless they are technical enough to know why its being presented.

I agree that I think it would be better if app developers could set the default to the optimal network config. If that is mac vlan now then it would be best if they could push users into that setup OOTB with users having to choose to go back to host network and ignore warnings for it.
